### ISSUE TITLE
Config generator emits uncompilable code for certain type conflicts

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#90976]
The implementation does name mangling, but those mangled names can conflict with other names, e.g.
```C#
using Microsoft.Extensions.Configuration;

IConfiguration configuration = new ConfigurationBuilder().Build();

var c1 = new Cint();
var c2 = new C<int>();
configuration.Bind(c1);
configuration.Bind(c2);

internal class C<T>
{
    public int Value { get; set; }
}

internal class Cint
{
    public int Value { get; set; }
}
```

**Comment by layomia** [comment]
I'll resolve with a shared abstraction doing whatever JSON does here.

**Comment by layomia** [comment]
Inlining this scenario from https://github.com/dotnet/runtime/issues/90969:

---

e.g.
```C#
namespace S1
{
    class C1
    {
        public void M(IConfiguration configuration)
        {
            var c = new C1();
            configuration.Bind(c);
        }
    }
}
namespace S2
{
    class C1
    {
        public void M(IConfiguration configuration)
        {
            var c = new C1();
            configuration.Bind(c);
        }
    }
}
```
results in lots of errors like:
```
error CS0111: Type 'BindingExtensions' already defines a member called 'Bind_C1' with the same parameter types
```

**Comment by ericstj** [comment]
Fixed in #91657
